storage/isolation_manager.rs

```diff
@@ -215,342 +346,319 @@ impl fmt::Display for IsolationError {
-        //
-        let (next_sequence_number, windows) = &*self.next_window_and_windows.read().unwrap_or_log();
-        windows
-            .iter()
-            .find_map(|w| {
-                if w.is_finished() {
-                    None
-                } else {
-                    debug_assert!(w.watermark().is_some());
-                    w.watermark()
-                }
-            })
-            .unwrap_or_else(|| SequenceNumber::new(next_sequence_number.number() - U80::new(1)))
+    fn may_free_windows(&self) {
+        let watermark = self.watermark();
+        let can_free_some: bool = {
+            let (next_relative_index, windows) = &*self.next_relative_index_and_windows.read().unwrap_or_log();
+            let start_of_second_window = *next_relative_index - ((windows.len() - 1) * TIMELINE_WINDOW_SIZE) as i64;
+            match windows.front() {
+                None => false,
+                Some(front) => front.get_readers() == 0 && watermark >= start_of_second_window,
+            }
+        };
+        if can_free_some {
+            let (next_relative_index, windows) = &mut *self.next_relative_index_and_windows.write().unwrap_or_log();
+            let mut end_of_first_remaining_window = *next_relative_index - ((windows.len() - 1) * TIMELINE_WINDOW_SIZE) as i64;
+            while watermark >= end_of_first_remaining_window && windows.front().is_some() && windows.front().unwrap().get_readers() == 0
+            {
+                windows.pop_front();
+                end_of_first_remaining_window += TIMELINE_WINDOW_SIZE as i64;
+            }
+        }
     }
 
-    fn record_reader(&self, sequence_number: SequenceNumber) {
-        let window = self.get_or_create_window(sequence_number);
-        window.increment_readers();
+    fn may_increment_watermark(&self, relative_index: i64) {
+        let mut watermark = self.watermark.load(Ordering::SeqCst);
+        if relative_index != watermark + 1 {
+            return ();
+        }
+
+        let (mut window, mut slot_index) = self.try_get_window(relative_index).unwrap();
+        let end_of_window: i64 = relative_index + (TIMELINE_WINDOW_SIZE - slot_index) as i64;
+        loop {
+            let should_update: bool = match window.get_status(slot_index) {
+                CommitStatus::Empty | CommitStatus::Pending(_, _) | CommitStatus::Validated(_, _) => false,
+                CommitStatus::Applied(_, _) => true,
+                CommitStatus::Aborted(_) => true,
+            };
+            if should_update
+                && self.watermark.compare_exchange(watermark, watermark + 1, Ordering::SeqCst, Ordering::SeqCst).is_ok()
+            {
+                watermark += 1;
+                slot_index += 1;
+                if slot_index >= TIMELINE_WINDOW_SIZE {
+                    if let Some(res) = self.try_get_window(watermark + 1) {
```

Comment:
If watermark and slot index move together, do we need to increment again?

Reply:
You can simplify this expression a bit:
```
if let Some((window, slot_index)) = self.try_get_window(...) {
  assert_eq(slot_index, 0);
} else {
  break;
}
```

Reply:
> If watermark and slot index move together, do we need to add to the watermark again?
`watermark` is the current watermark. slot_index is the slot index of the next candidate watermark  = `watermark+1`. 

> You can simplify this expression a bit:
`if let Some((window, slot_index))` seems to redeclare slot_index as a new variable, so this doesn't work.